### PR TITLE
feat: redesign cell store

### PR DESCRIPTION
**Pull request by zhangsoledad**
## Description
Previous cell storage is inefficient, resolve transactions need fetch cell from block body, this PR proposal a new live cell storage schema, optimize resolve transaction bottleneck, I run a benchmark (500 2-in-2out secp lock txs/per block), `process block` get 4x faster.
It precondition `chain freezer` which intends to achieve historical block.

## Breaking changes
remove `get_cells_by_lock_hash` rpc.

### CODE REVIEW

**Comment by zhangsoledad** [comment]
benchmark

**Comment by zhangsoledad** [comment]
```diff
diff --git a/chain/src/cell.rs b/store/src/cell.rs
index b2154f22d..be8426488 100644
--- a/chain/src/cell.rs
+++ b/store/src/cell.rs
@@ -1,88 +1,141 @@
+use crate::{ChainStore, StoreTransaction};
 use ckb_error::Error;
-use ckb_store::{ChainStore, StoreTransaction};
-use ckb_types::{
-    core::{BlockView, TransactionMeta},
-    prelude::*,
-};
+use ckb_types::{core::BlockView, packed, prelude::*};
+use std::collections::HashMap;
 
+/**
+ * Live cell entry.
+ *
+ *  table CellEntry {
+ *      output:                CellOutput,
+ *      block_hash:            Byte32,
+ *      block_number:          Uint64,
+ *      block_epoch:           Uint64,
+ *      index:                 Uint32,
+ *      data_size:             Uint64,
+ *  }
+ *
+ *
+ *  table CellDataEntry {
+ *      output_data:           Bytes,
+ *      output_data_hash:      Byte32,
+ *  }
+ */
+
+// Apply the effects of this block on the live cell set.
 pub fn attach_block_cell(txn: &StoreTransaction, block: &BlockView) -> Result<(), Error> {
-    for tx in block.transactions() {
-        for cell in tx.input_pts_iter() {
-            let cell_tx_hash = cell.tx_hash();
-            if let Some(mut meta) = txn.get_tx_meta(&cell_tx_hash) {
-                meta.set_dead(cell.index().unpack());
-                if meta.all_dead() {
-                    txn.delete_cell_set(&cell_tx_hash)?;
-                } else {
-                    txn.update_cell_set(&cell_tx_hash, &meta.pack())?;
-                }
-            }
-        }
-        let tx_hash = tx.hash();
-        let outputs_len = tx.outputs().len();
-        let meta = if tx.is_cellbase() {
-            TransactionMeta::new_cellbase(
-                block.number(),
-                block.epoch().number(),
-                block.hash(),
-                outputs_len,
-                false,
-            )
-        } else {
-            TransactionMeta::new(
-                block.number(),
-                block.epoch().number(),
-                block.hash(),
-                outputs_len,
-                false,
-            )
-        };
-        txn.update_cell_set(&tx_hash, &meta.pack())?;
-    }
+    let transactions = block.transactions();
+
+    // add new live cells
+    let new_cells = transactions
+        .iter()
+        .enumerate()
+        .map(move |(tx_index, tx)| {
+            let tx_hash = tx.hash();
+            let block_hash = block.header().hash();
+            let block_number = block.header().number();
+            let block_epoch = block.header().epoch();
+
+            tx.outputs_with_data_iter()
+                .enumerate()
+                .map(move |(index, (cell_output, data))| {
+                    let out_point = packed::OutPoint::new_builder()
+                        .tx_hash(tx_hash.clone())
+                        .index(index.pack())
+                        .build();
+                    let data_hash = packed::CellOutput::calc_data_hash(&data);
+
+                    let entry = packed::CellEntryBuilder::default()
+                        .output(cell_output)
+                        .block_hash(block_hash.clone())
+                        .block_number(block_number.pack())
+                        .block_epoch(block_epoch.pack())
+                        .index(tx_index.pack())
+                        .data_size((data.len() as u64).pack())
+                        .build();
+
+                    let data_entry = packed::CellDataEntryBuilder::default()
+                        .output_data(data.pack())
+                        .output_data_hash(data_hash)
+                        .build();
+
+                    (out_point, entry, data_entry)
+                })
+        })
+        .flatten();
+    txn.insert_cells(new_cells)?;
+
+    // mark inputs dead
+    // skip cellbase
+    let deads = transactions
+        .iter()
+        .skip(1)
+        .map(|tx| tx.input_pts_iter())
+        .flatten();
+    txn.delete_cells(deads)?;
+
     Ok(())
 }
 
+// Undo the effects of this block on the live cell set.
 pub fn detach_block_cell(txn: &StoreTransaction, block: &BlockView) -> Result<(), Error> {
-    for tx in block.transactions().iter().rev() {
-        txn.delete_cell_set(&tx.hash())?;
-
-        for cell in tx.input_pts_iter() {
-            let cell_tx_hash = cell.tx_hash();
-            let index: usize = cell.index().unpack();
-
-            if let Some(mut tx_meta) = txn.get_tx_meta(&cell_tx_hash) {
-                tx_meta.unset_dead(index);
-                txn.update_cell_set(&cell_tx_hash, &tx_meta.pack())?;
-            } else {
-                // the tx is full dead, deleted from cellset, we need recover it when fork
-                if let Some((tx, header)) =
-                    txn.get_transaction(&cell_tx_hash)
-                        .and_then(|(tx, block_hash)| {
-                            txn.get_block_header(&block_hash).map(|header| (tx, header))
-                        })
-                {
-                    let mut meta = if tx.is_cellbase() {
-                        TransactionMeta::new_cellbase(
-                            header.number(),
-                            header.epoch().number(),
-                            header.hash(),
-                            tx.outputs().len(),
-                            true, // init with all dead
-                        )
-                    } else {
-                        TransactionMeta::new(
-                            header.number(),
-                            header.epoch().number(),
-                            header.hash(),
-                            tx.outputs().len(),
-                            true, // init with all dead
-                        )
-                    };
-                    meta.unset_dead(index); // recover
-                    txn.update_cell_set(&cell_tx_hash, &meta.pack())?;
-                }
-            }
+    let transactions = block.transactions();
+    let mut input_pts = HashMap::with_capacity(transactions.len());
+
+    for tx in transactions.iter().skip(1) {
+        for pts in tx.input_pts_iter() {
+            let tx_hash = pts.tx_hash();
+            let index: usize = pts.index().unpack();
+            let indexes = input_pts.entry(tx_hash).or_insert_with(Vec::new);
+            indexes.push(index);
         }
     }
+
+    // restore inputs
+    // skip cellbase
+    let undo_deads = input_pts
+        .iter()
+        .filter_map(|(tx_hash, indexes)| {
+            txn.get_transaction_with_info(&tx_hash)
+                .map(move |(tx, info)| {
+                    let block_hash = info.block_hash;
+                    let block_number = info.block_number;
+                    let block_epoch = info.block_epoch;
+                    let tx_index = info.index;
+
+                    indexes.iter().filter_map(move |index| {
+                        tx.output_with_data(*index).map(|(cell_output, data)| {
+                            let out_point = packed::OutPoint::new_builder()
+                                .tx_hash(tx_hash.clone())
+                                .index(index.pack())
+                                .build();
+                            let data_hash = packed::CellOutput::calc_data_hash(&data);
+
+                            let entry = packed::CellEntryBuilder::default()
+                                .output(cell_output)
+                                .block_hash(block_hash.clone())
+                                .block_number(block_number.pack())
+                                .block_epoch(block_epoch.pack())
+                                .index(tx_index.pack())
+                                .data_size((data.len() as u64).pack())
+                                .build();
+
+                            let data_entry = packed::CellDataEntryBuilder::default()
+                                .output_data(data.pack())
+                                .output_data_hash(data_hash)
+                                .build();
+
+                            (out_point, entry, data_entry)
+                        })
+                    })
+                })
+        })
+        .flatten();
+    txn.insert_cells(undo_deads)?;
+
+    // undo live cells
+    let undo_cells = transactions.iter().map(|tx| tx.output_pts_iter()).flatten();
+    txn.delete_cells(undo_cells)?;
+
     Ok(())
 }
```

**Comment by zhangsoledad** [comment]
https://github.com/nervosnetwork/ckb/blob/cbc369b7f44b10721d27ede5c17bf7ec45e3ccd5/shared/src/migrations/cell.rs#L14-L56

**Comment by nervos-bot-user** [comment]
#### Benchmark Result
  - TPS: 478
  - CKB Version: cbc369b7f44b10721d27ede5c17bf7ec45e3ccd5
  - Instance Type: c5.xlarge
  - Instances Count: 3
  - Bench Type: In2Out2
  - CKB Logger Filter: info,ckb=debug

**Comment by chuijiaolianying** [comment]
Sync test executed in AWS with location is Singapore, from 0 to 2500000 height. 

Result here: (the unit is block_count/s)

  | avg | median | min | max | count
-- | -- | -- | -- | -- | --
redesign_cell_store | 578 | 569 | 485 | 676 | 52
v0.35.0 | 618 | 615 | 539 | 702 | 54

Original data: 
[redesign_cell_store.log](https://github.com/nervosnetwork/ckb/files/5215630/redesign_cell_store.log)
[v0.35.0_result.log](https://github.com/nervosnetwork/ckb/files/5215631/v0.35.0_result.log)

**Comment by zhangsoledad** [comment]
benchmark

**Comment by nervos-bot-user** [comment]
#### Benchmark Result
  - TPS: 495
  - CKB Version: b77391e01c5250009b7cc7ae3b65f58a4cdd2b3d
  - Instance Type: c5.xlarge
  - Instances Count: 3
  - Bench Type: In2Out2
  - CKB Logger Filter: info,ckb=debug

**Comment by doitian** [comment]
What's the relationship between this PR and #2203?

**Comment by chuijiaolianying** [comment]
Sync test executed in AWS with location is Singapore, from 0 to 2500000 height. 

Result here: (the unit is block_count/s)

  | avg | median | min | max | count
-- | -- | -- | -- | -- | --
redesign_cell_store | 587 | 578 | 531 | 665 | 21
commit_a2378b5 | 585 | 573 | 545 | 647 | 21

Original data: 
[ckb_cell_store.log](https://github.com/nervosnetwork/ckb/files/5222843/ckb_cell_store.log)
[ckb_a2378b5.log](https://github.com/nervosnetwork/ckb/files/5222844/ckb_a2378b5.log)

File: util/types/schemas/blockchain.mol [hunk 1]
```diff
@@ -108,3 +108,19 @@ table WitnessArgs {
     input_type:             BytesOpt,          // Type args for input
     output_type:            BytesOpt,          // Type args for output
 }
+
+/* Types for CellDB */
```
Comment (quake) [inline util/types/schemas/blockchain.mol:112]: suggest move these schema to `extensions.mol`, there is a section for `/* Types for Storage */`

**Review by quake — CHANGES_REQUESTED** [review]

File: store/src/cell.rs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+                    let entry = packed::CellEntryBuilder::default()
+                        .output(cell_output)
+                        .block_hash(block_hash.clone())
+                        .block_number(block_number.pack())
+                        .block_epoch(block_epoch.pack())
+                        .index(tx_index.pack())
+                        .data_size((data.len() as u64).pack())
+                        .build();
+
+                    let data_entry = packed::CellDataEntryBuilder::default()
```
[the remark this answers is not in the document]
Comment (zhangsoledad) [reply store/src/cell.rs:57]: The original storage interface like `fn get_cell_data(&'a self....-> Option<(Bytes, packed::Byte32)>`,
it distinguishes an empty data cell from a non-exist cell, I think it's a bad idea to mix them up.

and we precomputed blank hash
https://github.com/nervosnetwork/ckb/blob/52d9264145b101df9f7faf621fdea03a313306dd/util/hash/src/lib.rs#L17-L22
it's not so expensive.

[the remark this answers is not in the document]
Comment (quake) [reply store/src/cell.rs:57]: insert_cells fn inserts an empty cell_data slice (48 bytes):

```
self.insert_raw(COLUMN_CELL_DATA, &key, cell_data.as_slice())?;
```

my suggestion is skipping this empty molecule data, it reduces 48 bytes storage for each empty output:
```
if cell_data.is_none() {
    self.insert_raw(COLUMN_CELL_DATA, &key, [])?;
} else {
    ...
}
```

it won't change the interface of get_cell_data

[the remark this answers is not in the document]
Comment (zhangsoledad) [reply store/src/cell.rs:57]: You've missed my point, I think `get_cell_data` return means cell doesn't exist, not empty. 
So if we skip empty cell_data, how we distinguish non-exist from empty here?
it won't change the interface of get_cell_data, but it changes convention.
I don't think 48 bytes is matter. why we should mix them together?
like other storage API return None means non-exist, not empty.
https://explorer.nervos.org/charts/cell-count
assume all live-cell now are empty, it just 7MB, who care 7MB?

**Review by driftluo — APPROVED** [review]

**Review by quake — APPROVED** [review]

**Comment by quake** [comment]
bors r=quake,driftluo

File: util/types/schemas/extensions.mol [hunk 3]
```diff
@@ -70,13 +70,18 @@ struct TransactionInfo {
...
-    epoch_number:   Uint64,
-    len:            Uint32,
-    bits:           Bytes,
-    cellbase:       Bool,
+table CellEntry {
+    output:                CellOutput,
+    block_hash:            Byte32,
+    block_number:          Uint64,
+    block_epoch:           Uint64,
+    index:                 Uint32,
```
Comment (doitian) [inline util/types/schemas/extensions.mol:78]: `tx_index` is more verbose.

```suggestion
    tx_index:                 Uint32,
```

**Pull request merged**